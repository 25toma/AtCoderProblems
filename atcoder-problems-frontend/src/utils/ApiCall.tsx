import * as request from "superagent";
import { Problem } from "../model/Problem";
import { Contest } from "../model/Contest";
import { Submission } from "../model/Submission";
import { RankPair } from "../model/RankPair";
import { MergedProblem } from "../model/MergedProblem";
import { RankingKind } from "../model/RankingKind";

export class ApiCall {
  static BaseUrl = "./atcoder-api";

  static getJson(url: string, query?: any): Promise<any> {
    return new Promise((resolve, reject) => {
      request
        .get(url)
        .query(query)
        .set("Content-Type", "application/json")
        .end((err, res) => {
          if (err) {
            reject(err);
          } else {
            resolve(res.body);
          }
        });
    });
  }

  static getProblems(): Promise<Array<Problem>> {
    let url = `${this.BaseUrl}/info/problems`;
    return this.getJson(url).then((obj: Array<any>) => {
      let problems: Problem[] = obj.map(o => {
        let p = { id: o["id"], title: o["title"], contestId: o["contest_id"] };
        return p;
      });
      return problems;
    });
  }

  static getRanking(kind: string): Promise<Array<RankPair>> {
    let url = `${this.BaseUrl}/info/${kind}`;
    return this.getJson(url).then((obj: Array<any>) => {
      let ranks = obj.map(o => {
        let p = { rank: 1, userId: o["user_id"], count: o["problem_count"] };
        return p;
      });
      ranks.sort((a, b) => b.count - a.count);

      for (let i = 1; i < ranks.length; i += 1) {
        if (ranks[i - 1].count == ranks[i].count) {
          ranks[i].rank = ranks[i - 1].rank;
        } else {
          ranks[i].rank = i + 1;
        }
      }

      return ranks;
    });
  }

  static getContests(): Promise<Array<Contest>> {
    let url = `${this.BaseUrl}/info/contests`;
    return this.getJson(url).then((obj: Array<any>) => {
      let contests: Contest[] = obj.map(o => {
        return {
          id: o["id"],
          title: o["title"],
          start_epoch_second: o["start_epoch_second"]
        };
      });
      return contests;
    });
  }

  static getSubmissions(
    url: string,
    query?: { user: string; rivals: string }
  ): Promise<Array<Submission>> {
    return this.getJson(url, query).then((obj: Array<any>) => {
      let submissions: Submission[] = obj.map(o => {
        return {
          point: o["point"],
          result: o["result"],
          problem_id: o["problem_id"],
          user_id: o["user_id"],
          epoch_second: o["epoch_second"],
          id: o["id"],
          language: o["language"],
          length: o["length"],
          contestId: o["contest_id"],
          execution_time: o["execution_time"]
        };
      });
      return submissions;
    });
  }

  static getMergedProblems(): Promise<Array<MergedProblem>> {
    let url = `${this.BaseUrl}/info/merged-problems`;
    return this.getJson(url).then((obj: Array<any>) => {
      let problems: MergedProblem[] = obj.map(o => {
        return {
          first_submission_id: o["first_submission_id"],
          solver_count: o["solver_count"],
          fastest_user_id: o["fastest_user_id"],
          execution_time: o["execution_time"],
          shortest_user_id: o["shortest_user_id"],
          shortest_submission_id: o["shortest_submission_id"],
          contestId: o["contest_id"],
          id: o["id"],
          fastest_submission_id: o["fastest_submission_id"],
          first_user_id: o["first_user_id"],
          title: o["title"],
          source_code_length: o["source_code_length"],
          fastest_contest_id: o["fastest_contest_idw"],
          shortest_contest_id: o["shortest_contest_id"],
          first_contest_id: o["first_contest_id"],
          point: o["point"],
          predict: o["predict"]
        };
      });
      return problems;
    });
  }
}
