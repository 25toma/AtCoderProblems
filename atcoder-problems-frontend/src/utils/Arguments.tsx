import * as QueryString from "query-string";
import { RankingKind } from "../components/Ranking";

export interface Arguments {
  userId: string;
  rivals: Array<string>;
  kind: string;
  ranking: string;
}

export class ArgumentParser {
  static parse(): Arguments {
    let params = QueryString.parse(location.search);
    let userId: string = "user" in params ? params["user"] : "";
    let rivals: string = "rivals" in params ? params["rivals"] : "";
    let kind: string = "kind" in params ? params["kind"] : "category";
    let ranking: string = "ranking" in params ? params["ranking"] : "ac";
    return {
      userId: userId,
      rivals: rivals.split(","),
      kind: kind,
      ranking: ranking
    };
  }
}
