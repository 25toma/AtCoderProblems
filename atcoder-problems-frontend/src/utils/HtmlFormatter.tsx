import * as React from "react";

export class HtmlFormatter {
  /**
   * create hyperlink from url and link text
   * @param url reference url
   * @param text link text
   */
  static createLink(url: string, text: string): JSX.Element {
    return (
      <a href={url} target="_blank">
        {text}
      </a>
    );
  }
}
