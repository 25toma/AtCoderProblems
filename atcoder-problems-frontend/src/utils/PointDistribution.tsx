import { Submission } from "../model/Submission";
import { MergedProblem } from "../model/MergedProblem";

export class PointDistribution {
  /**
   * count accepted problems by point
   * @param submissions submissions
   * @param problems problem list
   * @param userId user id to filter
   * @returns Map(point -> count)
   */
  static countAcceptedByPoint(
    submissions: Array<Submission>,
    problems: Array<MergedProblem>,
    userId: String
  ): Map<number, number> {
    let problemSet = new Set(
      submissions
        .filter(s => s.result === "AC" && s.user_id === userId)
        .map(s => s.problem_id)
    );

    let count = new Map<number, number>();
    problems.filter(p => problemSet.has(p.id) && p.point).forEach(p => {
      if (!count.has(p.point)) {
        count.set(p.point, 0);
      }
      count.set(p.point, count.get(p.point) + 1);
    });

    return count;
  }
}
