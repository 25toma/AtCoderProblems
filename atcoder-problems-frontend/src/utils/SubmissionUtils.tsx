import { Submission } from "../model/Submission";

export class SubmissionUtils {
  /**
   * extract problem ids of the given result with the given users
   * @param submissions list of submissions
   * @param userIds set of user ids
   * @param resultToExtract set of results to extract
   * @returns map of <problem id, submission result>
   */
  static extractProblemIdsByUsers(
    submissions: Array<Submission>,
    userIds: Set<string>,
    resultToExtract: Set<string> = new Set(["AC"])
  ): Map<string, string> {
    let problemIdSet = new Set<string>();
    let problemPairs: [string, string][] = submissions
      .filter(s => resultToExtract.has(s.result) && userIds.has(s.user_id))
      .map(s => {
        let t: [string, string] = [s.problem_id, s.result];
        return t;
      });

    return new Map(problemPairs);
  }
}
