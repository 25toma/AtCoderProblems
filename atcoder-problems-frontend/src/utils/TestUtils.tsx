export class MockRequest {
  constructor(body: any, error?: any) {
    this.body = body;
    this.error = error;
  }
  get() {
    return this;
  }
  set() {
    return this;
  }
  query() {
    return this;
  }
  body: any = null;
  error: any = null;
  end = jest.fn().mockImplementation(callback => {
    callback(this.error, {
      status() {
        return 200;
      },
      ok() {
        return true;
      },
      get: jest.fn(),
      body: this.body,
      toError: jest.fn()
    });
  });
}
