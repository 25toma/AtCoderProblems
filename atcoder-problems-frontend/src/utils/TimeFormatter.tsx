import * as moment from "moment";

export class TimeFormatter {
  static getDateString(currentTimeMillis: number): string {
    let date = moment(currentTimeMillis);
    return date.local().format("Y-MM-DD");
  }
}
