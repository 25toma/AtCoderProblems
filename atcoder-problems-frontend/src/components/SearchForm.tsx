import * as React from "react";
import { Arguments } from "../utils/Arguments";
import { FormGroup, ControlLabel, FormControl, Radio, Button, Form, ToggleButton, ToggleButtonGroup } from "react-bootstrap";

export interface SearchFormProps {
    args: Arguments;
}

export class SearchForm extends React.Component<SearchFormProps, {}>{
    render() {
        let userId = this.props.args.userId;
        let rivals = this.props.args.rivals.join(",");
        let kind = this.props.args.kind;
        return (
            <Form action="./" method="get" inline>
                <FormGroup>
                    <ControlLabel>User ID: </ControlLabel>
                    <FormControl placeholder="UserID" type="text" name="user" defaultValue={userId} />
                </FormGroup>
                {' '}
                <FormGroup>
                    <ControlLabel>Rivals: </ControlLabel>
                    <FormControl placeholder="UserID,UserID,..." type="text" name="rivals" defaultValue={rivals} />
                </FormGroup>
                {' '}
                <ToggleButtonGroup type="radio" name="kind" defaultValue={kind}>
                    <ToggleButton value="category">Category</ToggleButton>
                    <ToggleButton value="list">List</ToggleButton>
                </ToggleButtonGroup>
                {' '}
                <Button type="submit">Search</Button>
            </Form>
        );
    }
}
