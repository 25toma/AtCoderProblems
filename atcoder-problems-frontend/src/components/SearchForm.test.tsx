import * as fs from "fs";
import * as React from "react";
import { SearchForm } from "./SearchForm";
import { shallow, mount } from "enzyme";
import { configure } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import { SubmissionUtils } from "../utils/SubmissionUtils";

configure({ adapter: new Adapter() });
test("render list view", () => {
  let wrapper = mount(
    <SearchForm
      args={{
        userId: "kenkoooo",
        rivals: ["chokudai", "iwiwi"],
        kind: "kind",
        ranking: ""
      }}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
