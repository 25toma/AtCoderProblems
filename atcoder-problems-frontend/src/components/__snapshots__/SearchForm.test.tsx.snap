// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render list view 1`] = `
<SearchForm
  args={
    Object {
      "kind": "kind",
      "rivals": Array [
        "chokudai",
        "iwiwi",
      ],
      "userId": "kenkoooo",
    }
  }
>
  <Form
    action="./"
    bsClass="form"
    componentClass="form"
    horizontal={false}
    inline={true}
    method="get"
  >
    <form
      action="./"
      className="form-inline"
      method="get"
    >
      <FormGroup
        bsClass="form-group"
      >
        <div
          className="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            <label
              className="control-label"
            >
              User ID: 
            </label>
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            componentClass="input"
            defaultValue="kenkoooo"
            name="user"
            placeholder="UserID"
            type="text"
          >
            <input
              className="form-control"
              defaultValue="kenkoooo"
              name="user"
              placeholder="UserID"
              type="text"
            />
          </FormControl>
        </div>
      </FormGroup>
       
      <FormGroup
        bsClass="form-group"
      >
        <div
          className="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            <label
              className="control-label"
            >
              Rivals: 
            </label>
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            componentClass="input"
            defaultValue="chokudai,iwiwi"
            name="rivals"
            placeholder="UserID,UserID,..."
            type="text"
          >
            <input
              className="form-control"
              defaultValue="chokudai,iwiwi"
              name="rivals"
              placeholder="UserID,UserID,..."
              type="text"
            />
          </FormControl>
        </div>
      </FormGroup>
       
      <Uncontrolled(ToggleButtonGroup)
        defaultValue="kind"
        name="kind"
        type="radio"
      >
        <ToggleButtonGroup
          name="kind"
          onChange={[Function]}
          type="radio"
          value="kind"
        >
          <ButtonGroup
            block={false}
            bsClass="btn-group"
            data-toggle="buttons"
            justified={false}
            vertical={false}
          >
            <div
              className="btn-group"
              data-toggle="buttons"
            >
              <ToggleButton
                checked={false}
                key=".0"
                name="kind"
                onChange={[Function]}
                type="radio"
                value="category"
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  componentClass="label"
                  disabled={false}
                >
                  <label
                    className="btn btn-default"
                    disabled={false}
                    type="button"
                  >
                    <input
                      autoComplete="off"
                      checked={false}
                      disabled={false}
                      name="kind"
                      onChange={[Function]}
                      type="radio"
                      value="category"
                    />
                    Category
                  </label>
                </Button>
              </ToggleButton>
              <ToggleButton
                checked={false}
                key=".1"
                name="kind"
                onChange={[Function]}
                type="radio"
                value="list"
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  componentClass="label"
                  disabled={false}
                >
                  <label
                    className="btn btn-default"
                    disabled={false}
                    type="button"
                  >
                    <input
                      autoComplete="off"
                      checked={false}
                      disabled={false}
                      name="kind"
                      onChange={[Function]}
                      type="radio"
                      value="list"
                    />
                    List
                  </label>
                </Button>
              </ToggleButton>
            </div>
          </ButtonGroup>
        </ToggleButtonGroup>
      </Uncontrolled(ToggleButtonGroup)>
       
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        type="submit"
      >
        <button
          className="btn btn-default"
          disabled={false}
          type="submit"
        >
          Search
        </button>
      </Button>
    </form>
  </Form>
</SearchForm>
`;
